File: src/empirical_vs_ml/site_specific/gen_emp_perf_table.jl
include("../empirical.jl")

using MLJ
using CSV
using DataFrames
using Statistics

using ..EmpiricalModels

"""
Implementa o cáculo do Mean Bias Error
"""
function mbe(preds::AbstractVector, real::AbstractVector)
    return sum((real .- preds)) / length(preds)
end



"""
Implementa o cáculo do R2 Score 
"""
function r2(preds::AbstractVector, real::AbstractVector)

    real_mean = mean(real)

    var_Y_hat = sum((preds .- real_mean).^2)
    var_Y_bar = sum((real .- real_mean).^2)

    return (var_Y_bar - var_Y_hat) / var_Y_bar
end


function main()

    if !isdir("./empirical_preds_julia/")
        mkdir("./empirical_preds_julia/")
    end
    
    emp_perf_compare = DataFrame(station = String[],
                                 rmse = Float64[],
                                 mae = Float64[],
                                 mbe = Float64[],
                                 r2 = Float64[])
       
    for file in readdir("./input_data")

        input_data = CSV.read("./input_data/$file", DataFrame)

        if isempty(input_data)
            continue
        end

        println("processando $file...")

        this_site_preds = DataFrame(doy = Int64[],
                                    hour = Int64[],
                                    year = Int64[],
                                    emp_pred = Float64[],
                                    real = Float64[])


        lat = input_data[begin, :lat]
        lon = input_data[begin, :lon]

        # gera as previsões e as guara em conjuntos de previsoes
        for row in eachrow(input_data)

            localtime = row[:hour] - 3
            standard_meridian_time = -3
            doy = row[:doy]
            total_daily = row[:daily_gsr]

            emp_pred = EmpiricalModels.predict(lat, lon, localtime, standard_meridian_time, doy, total_daily)

            info = (doy, localtime, row[:year], emp_pred, row[:radiacao_global])
            push!(this_site_preds, info)

        end

        CSV.write("./empirical_preds_julia/$file" , this_site_preds)

        pred = this_site_preds[:, :emp_pred]
        real = this_site_preds[:, :real]

        perf_info = (split(file, ".")[begin],
                     rmse(pred, real),
                     mae(pred, real),
                     mbe(pred, real),
                     r2(pred, real) )


        push!(emp_perf_compare, perf_info)
    
    end

    CSV.write("./empirical_performance_julia.csv" , emp_perf_compare)

end

main()

    

File: src/empirical_vs_ml/site_specific/gen_emp_perf_table.py
import os
import numpy as np
import pandas as pd

from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

from empirical_vs_ml.empirical_model import CPRGModel


def mean_bias_error(preds, real) -> float:

    preds = np.asarray(preds)
    real = np.asarray(real)

    # preds é unidimensional então não e possível subscrever o .size
    return np.sum(preds - real) / preds.size


if __name__ == "__main__":

    perf_info = {
        'station': [],
        'lat': [],
        'lon': [],
        'rmse': [],
        'mae': [],
        'mbe': [],
        'r2': [],
    }

    if not os.path.exists('./empirical_preds_python/'):
        os.mkdir('./empirical_preds_python/')


    model = CPRGModel()

    for file in os.listdir("../input_data"):

        data = pd.read_csv(f"../input_data/{file}")
        station = file.split(".")[0]

        # pula os datamframes vazios
        if data.empty:
            continue

        this_site_preds = {
            'doy': [],
            'hour': [],
            'year': [],
            'emp_pred': [],
            'real': [],
        }

        lat = data.loc[0, 'lat']
        lon = data.loc[0, 'lon']

        for _, row in data.iterrows():

            localtime = row['hour'] - 3 # correção para o fuso-horário 
            standard_meridian_time = -3
            doy = row['doy']
            total_daily = row['daily_gsr']

            # localtime - 3 pois as previsões pareciam estar deslocadas em 3 horas mesmo fazendo
            # a correção para o fuso horário local. Com esse valor os erros são mehores
            emp_pred = model.predict(lat, lon, localtime - 3, standard_meridian_time, doy, total_daily)

            this_site_preds['doy'].append( row['doy'] )
            this_site_preds['hour'].append( row['hour'] )
            this_site_preds['year'].append( row['year'] )
            this_site_preds['emp_pred'].append( emp_pred )
            this_site_preds['real'].append( row['radiacao_global'] )


        this_site_preds = pd.DataFrame().from_dict( this_site_preds )
        this_site_preds.to_csv(f'./empirical_preds_python/{station}.csv', index = False)

        perf_info['station'].append( station )
        perf_info['lat'].append( lat )
        perf_info['lon'].append( lon )
        perf_info['rmse'].append( mean_squared_error(this_site_preds['emp_pred'], this_site_preds['real'], squared = False) )
        perf_info['mae'].append( mean_absolute_error(this_site_preds['emp_pred'], this_site_preds['real']) )
        perf_info['mbe'].append( mean_bias_error(this_site_preds['emp_pred'], this_site_preds['real']) )
        perf_info['r2'].append( r2_score(this_site_preds['emp_pred'], this_site_preds['real']) )


    perf_info = pd.DataFrame().from_dict(perf_info)
    perf_info.to_csv('./emp_perf_table.csv', index = False)

   # latex_table_format(perf_info, output_file = 'data.tex')
